src/riscv/lib/src/jit/builder/instr_map.rs

```diff
@@ -0,0 +1,233 @@
+            .collect();
+        InstrMap { instructions }
+    }
+}
+
+impl<V> Index<InstrId> for InstrMap<V> {
+    type Output = V;
+
+    fn index(&self, index: InstrId) -> &Self::Output {
+        &self.instructions[index.0]
+    }
+}
+
+impl<V> IndexMut<InstrId> for InstrMap<V> {
+    fn index_mut(&mut self, index: InstrId) -> &mut Self::Output {
+        &mut self.instructions[index.0]
+    }
+}
+
+/// Builder for initialising an ([`AddrMap`], [`InstrMap`]) pair.
+pub(crate) struct InstrMapBuilder<T> {
+    addresses: Vec<Address>,
+    instructions: Vec<T>,
+}
+
+impl<T> InstrMapBuilder<T> {
+    /// Create a new builder with the specified capacity.
+    pub(crate) fn with_capacity(capacity: usize) -> Self {
+        Self {
+            addresses: Vec::with_capacity(capacity),
+            instructions: Vec::with_capacity(capacity),
+        }
+    }
+
+    /// Insert a new address and instruction pair to the builder. If the address
+    /// already exists, override the existing instruction with the new instruction.
+    pub(crate) fn insert(&mut self, address: Address, instruction: T) {
+        // insert the address and instruction at the same index by binary searching the address list
+        match self.addresses.binary_search(&address) {
+            Ok(idx) => {
+                self.instructions[idx] = instruction;
+            }
+            Err(idx) => {
+                self.addresses.insert(idx, address);
+                self.instructions.insert(idx, instruction);
+            }
+        }
+    }
+
+    /// Finalise the builder and produce the (`AddrMap`, `InstrMap`) pair.
+    pub(crate) fn build(self) -> (AddrMap, InstrMap<T>) {
+        let instr_map = InstrMap {
+            instructions: self.instructions,
+        };
+        let addr_map = AddrMap {
+            addresses: self.addresses,
+        };
+
+        (addr_map, instr_map)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use proptest::proptest;
+
+    use super::*;
+
+    proptest! {
+        #[test]
+        fn addr_map_translate_and_address_of_are_inverses(addresses in proptest::collection::vec(0u64..1_000, 1..50)) {
+            let mut sorted_addresses = addresses.clone();
+            sorted_addresses.sort_unstable();
+            sorted_addresses.dedup();
+
+            let addr_map = AddrMap { addresses: sorted_addresses.clone() };
+
+            for (idx, &address) in sorted_addresses.iter().enumerate() {
+                let instr_id = addr_map.translate(address).expect("Address should translate to an InstrId");
+                assert_eq!(instr_id.0, idx);
+                let retrieved_address = addr_map.address_of(instr_id);
+                assert_eq!(retrieved_address, address);
+            }
+        }
+
+        #[test]
+        fn instr_map_iterators_and_map_function_work(instructions in proptest::collection::vec(0u32..1_000, 1..50)) {
+            let instr_map = InstrMap { instructions: instructions.clone() };
```

Comment:
Same here, let's not bake assumptions into how `InstrMap` is created into this test. `InstrMapBuilder::build` is the way to produce a `InstrMap`, we should be mindful of that and not test other ways to create this data structure.